File: TicTacToe.java
import java.util.Scanner;

public class TicTacToe {
    static void printBoard(char[][] a) {
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                System.out.print(a[i][j] + " ");
            }
            System.out.println();
        }
    }

    static boolean checkWinner(char[][] a) {
        for (int i = 0; i < 3; i++) {
            if (a[i][0] != '*' && a[i][0] == a[i][1] && a[i][1] == a[i][2]) return true;
            if (a[0][i] != '*' && a[0][i] == a[1][i] && a[1][i] == a[2][i]) return true;
        }
        if (a[0][0] != '*' && a[0][0] == a[1][1] && a[1][1] == a[2][2]) return true;
        if (a[0][2] != '*' && a[0][2] == a[1][1] && a[1][1] == a[2][0]) return true;
        return false;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        char[][] a = new char[3][3];
        String ch1, ch2;
        char mood;

        System.out.println("**** WELCOME TO TIC TAC TOE ****");

        do {
            for (int i = 0; i < 3; i++)
                for (int j = 0; j < 3; j++)
                    a[i][j] = '*';

            System.out.print("First player name: ");
            ch1 = sc.next();

            System.out.print("Second player name: ");
            ch2 = sc.next();

            int turn;
            boolean winnerFound = false;

            for (turn = 0; turn < 9; turn++) {
                printBoard(a);
                int r = -1, c = -1;
                boolean valid = false;

                while (!valid) {
                    try {
                        if (turn % 2 == 0)
                            System.out.print(ch1 + ", enter row and col (1-3): ");
                        else
                            System.out.print(ch2 + ", enter row and col (1-3): ");

                        r = sc.nextInt();
                        c = sc.nextInt();

                        if (r < 1 || r > 3 || c < 1 || c > 3) {
                            System.out.println("Invalid range! Enter numbers between 1 and 3.");
                            continue;
                        }
                        if (a[r - 1][c - 1] != '*') {
                            System.out.println("That spot is already filled. Try again.");
                            continue;
                        }

                        valid = true;
                    } catch (Exception e) {
                        System.out.println("Invalid input! Please enter numbers only.");
                        sc.nextLine();
                    }
                }

                a[r - 1][c - 1] = (turn % 2 == 0) ? 'X' : 'O';

                if (checkWinner(a)) {
                    printBoard(a);
                    System.out.println((turn % 2 == 0 ? ch1 : ch2) + " wins!");
                    winnerFound = true;
                    break;
                }
            }

            if (!winnerFound)
                System.out.println("Draw!");

            System.out.print("Play again? (N/n to play again): ");
            mood = sc.next().charAt(0);

        } while (mood == 'n' || mood == 'N');

        sc.close();
        System.out.println("Game over. Thanks for playing!");
    }
}

File: tictactoe.cpp
#include <stdio.h>

void printBoard(char a[3][3]) {
    for (int i=0;i<3;i++) {
        for (int j=0;j<3;j++)
            printf("%c ", a[i][j]);
        printf("\n");
    }
}

int checkWinner(char a[3][3]) {
    for (int i=0;i<3;i++) {
        if (a[i][0] != '*' && a[i][0] == a[i][1] && a[i][1] == a[i][2]) return 1;
        if (a[0][i] != '*' && a[0][i] == a[1][i] && a[1][i] == a[2][i]) return 1;
    }
    if (a[0][0] != '*' && a[0][0] == a[1][1] && a[1][1] == a[2][2]) return 1;
    if (a[0][2] != '*' && a[0][2] == a[1][1] && a[1][1] == a[2][0]) return 1;
    return 0;
}

int main() {
    char a[3][3], ch1[15], ch2[15], mood;
    printf("****WELCOME TO TIC TAC TOE****\n");

    do {
        for (int i=0;i<3;i++)
            for (int j=0;j<3;j++)
                a[i][j] = '*';

        printf("First player name: ");
        scanf("%s", ch1);
        printf("Second player name: ");
        scanf("%s", ch2);

        int turn;
        for (turn=0; turn<9; turn++) {
            printBoard(a);
            int r, c;
            while (1) {
                if (turn % 2 == 0)
                printf("%s enter row and col (1-3): ", ch1);
                else
                printf("%s enter row and col (1-3): ", ch2);
                
                int result = scanf("%d %d", &r, &c);
                if (result != 2) {
                    printf("Invalid input! Please enter numbers.\n");
                    while (getchar() != '\n');
                    continue;
                }
                if (r < 1 || r > 3 || c < 1 || c > 3) {
                    printf("Invalid range! Enter between 1 and 3.\n");
                    continue;
                }
                if (a[r-1][c-1] != '*') {
                    printf("That spot is already filled. Try again.\n");
                    continue;
                }
                break;
            }

            if (r<1 || r>3 || c<1 || c>3 || a[r-1][c-1] != '*') {
                printf("Invalid move, try again.\n");
                turn--;
                continue;
            }

            a[r-1][c-1] = (turn % 2 == 0) ? 'X' : 'O';

            if (checkWinner(a)) {
                printBoard(a);
                printf("%s wins!\n", (turn % 2 == 0) ? ch1 : ch2);
                break;
            }
        }

        if (turn == 9)
            printf("Draw!\n");

        printf("Play again? (N/n to play again): ");
        scanf(" %c", &mood);
    } while (mood == 'n' || mood == 'N');

    return 0;
}
